src/distributions/geometric.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2018 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! The Geometric distribution.
+
+use Rng;
+use distributions::Distribution;
+
+use std::f64;
+
+/// The Geometric distribution.
+///
+/// This distribution is the number of failures in Bernoulli trials
+/// before one success.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::distributions::{Geometric, Distribution};
+///
+/// let d = Geometric::new(0.3);
+/// let v = d.sample(&mut rand::thread_rng());
+/// println!("{} is from a Geometric distribution", v);
+/// ```
+///
+#[derive(Clone, Copy, Debug)]
+pub struct Geometric {
+    /// Probability of success
+    prob: f64,
+}
+
+// To sample from the Geometric distribution we simply sample from a uniform
+// distribution and apply "floor (log(q) / log(1 - p))".
+//
+// If `p == 1.0`, simply return 0, as we will always succeed on the first trial.
+//
+
+impl Geometric {
+    /// Construct a new `Geometric` with the given probability of success `p`.
+    ///
+    /// For `p = 1.0`, the resulting distribution will always generate 0.
+    ///
+    /// # Panics
+    ///
+    /// If `p <= 0` or `p > 1`.
+    ///
+    #[inline]
+    pub fn new(p: f64) -> Geometric {
+        if p <= 0.0 || p > 1.0 {
+            panic!("Geometric::new not called with 0.0 < p <= 1.0");
+        }
+        Geometric { prob: p }
+    }
+}
+
+impl Distribution<u64> for Geometric {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u64 {
+        // Make sure to always return 0 for p = 1.0.
+        if self.prob == 1.0 { return 0; }
+        let q: f64 = rng.gen();
+        (q.ln() / (1.0 - self.prob).ln()).floor() as u64
```

Comment:
Compare with statrs: https://github.com/boxtown/statrs/blob/master/src/distribution/geometric.rs

By my calculations, we want one less than the ceiling of the ratio of the logs, which is one less than @boxtown's version (presumably his version is the number of trials to success, though it doesn't say, despite linking to an article which mentions the two possible versions right at the top).

Additionally, we should be sampling `x` from `[0, 1)` which implies `q` from `(0, 1]`; i.e. `q = OpenClosed01::sample(&mut rng);`. Helpfully this avoids `log(0)`.

There's one other question: which return type is most appropriate? I don't know the answer to that one, but don't see any issue with either `u64` or `f64`. (It's possible to implement both. Since `f64` is used for the calculations, perhaps that makes more sense.)